### PR TITLE
tests: Add an integ lib test for adding static IPv4

### PR DESCRIPTION
**Pull request by EdDev**
Configuring a static IPv4 address on an interface.

The assertion is embedded in the transaction verification,
which is on by default.

### CODE REVIEW

**Comment by tyll** [comment]
This needs #87 and #92 to be merged

File: tests/integration/apply_ip_interface_test.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        ]
+    }
+
+
+def test_add_static_ipv4(eth1_state):
+    eth1_state['interfaces'][0]['ipv4']['enabled'] = True
+    eth1_state['interfaces'][0]['ipv4']['address'] = [
+        {'ip': '192.168.122.250', 'prefix-length': 24}
+    ]
+    netapplier.apply(eth1_state)
```
Comment (tyll) [inline tests/integration/apply_ip_interface_test.py]: IMHO some kind of assertion should be added after applying the state.

Comment (EdDev) [reply tests/integration/apply_ip_interface_test.py]: @tyll , I agree. I am waiting for the dependencies to get merged to make it easier for me to work on this.

**Comment by EdDev** [comment]
ci test please

**Comment by EdDev** [comment]
ci test please

**Comment by EdDev** [comment]
Depends on #100 and #101 
It also requires another fix for IPv6 setting and reporting which seems to be faulty based on this test results.

**Comment by EdDev** [comment]
Also depends now on #102

File: tests/integration/testlib/statelib.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import netinfo
+
+
+INTERFACES = 'interfaces'
```
Comment (tyll) [inline tests/integration/testlib/statelib.py]: Shouldn't this constant be part of libnmstate?

File: tests/integration/static_ip_address_test.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+from libnmstate import netapplier
+
+from .testlib import assertlib
+from .testlib import statelib
+
+
+def test_add_static_ipv4():
+    desired_state = statelib.show_only(('eth1',))
+    eth1_desired_state = desired_state['interfaces'][0]
```
Comment (tyll) [inline tests/integration/static_ip_address_test.py]: This is not using `INTERFACES`... is this intended?

File: tests/integration/testlib/statelib.py (hunk 2 above)
Comment (EdDev) [reply tests/integration/testlib/statelib.py]: I guess it should be promoted there. A PR on that will be nice.
(where to locate it is less obvious I guess)

File: tests/integration/static_ip_address_test.py (hunk 3 above)
Comment (EdDev) [reply tests/integration/static_ip_address_test.py]: Thanks for noticing, changing.

File: tests/integration/testlib/statelib.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        INTERFACES: [{'name': ifname} for ifname in ifnames]
+    }
+    return filtered_current_state(base_filter_state)
+
+
+def filtered_current_state(desired_state):
+    """
+    Given the desired state, return a filtered version of the current state
+    which includes only entities that have been mentioned in the desired state.
+    In case there are no interfaces for filtering, all are reported
```
Comment (tyll) [inline tests/integration/testlib/statelib.py]: The doctext mentions entities without explaining what they are and used interfaces in the next sentence, maybe use `entities such as interfaces` and then `entities` in the second sentence.

File: tests/integration/testlib/assertlib.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+def assert_state(desired_state):
+    """Given a state, assert it against the current state."""
+    filtered_current_state = statelib.filtered_current_state(desired_state)
+    normalized_desired_state = _normalize_desired_state(desired_state,
+                                                        filtered_current_state)
+
+    assert normalized_desired_state == filtered_current_state
+
+
+def _normalize_desired_state(desired_state, current_state):
```
Comment (tyll) [inline tests/integration/testlib/assertlib.py]: Isn't something like this part of libnmstate?

File: tests/integration/testlib/statelib.py (hunk 4 above)
Comment (EdDev) [reply tests/integration/testlib/statelib.py]: Done

Comment (tyll) [reply tests/integration/testlib/statelib.py]: Also IMHO filtered is to unpecific for the function. Since it works on entities, this should be part of the name.

File: tests/integration/testlib/assertlib.py (hunk 5 above)
Comment (EdDev) [reply tests/integration/testlib/assertlib.py]: We want the test to assert the outcome even though the production code is already doing that (you pointed that out well in the last review).
This is the test version of such an assert which require normalization of the data.
Hopefully, we will not repeat bugs in production here, or at least have different bugs.

I also hope we will have less logic here, as we do not need the inner metadata and other special handling, we expect nmstate to return back something that is simpler.

File: tests/integration/static_ip_address_test.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+from libnmstate import netapplier
+
+from .testlib import assertlib
+from .testlib import statelib
+from .testlib.statelib import INTERFACES
+
+
+def test_add_static_ipv4():
+    desired_state = statelib.show_only(('eth1',))
```
Comment (tyll) [inline tests/integration/static_ip_address_test.py]: IMHO the test should work without getting the current state of eth1 but specify the desired state indepentently and then check if it is properly applied. Otherwise the test condition depends on the previous state of the test setup which seems to be wrong.

File: tests/integration/testlib/statelib.py (hunk 4 above)
Comment (EdDev) [reply tests/integration/testlib/statelib.py]: Could you please explain what you mean? It should probably be a verb: `filter_current_state`. And it works on a state, I think it is obvious the state has entities.

File: tests/integration/static_ip_address_test.py (hunk 6 above)
Comment (EdDev) [reply tests/integration/static_ip_address_test.py]: I think this is safer as it does not assume anything about what exists now.
My main concern is when we add other properties which may be a must: I do not want to revisit the test. In this solution, we get the template from what is supported now by nmstate and edit it.

We may want to add tests that check specific completion of data (when we provide only minimum data and expect nmstate to merge it with current), but I think that is not the same test... at least, it is a more messy one.

I agree we need not to drag problems or states between tests. For that I plan another change to be executed on every executed test, I will post it after we finish this initial one.

File: tests/integration/testlib/assertlib.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import collections
+import copy
+import six
+
+from . import statelib
+from .statelib import INTERFACES
+
+
+def assert_state(desired_state):
+    """Given a state, assert it against the current state."""
```
Comment (cathay4t) [inline tests/integration/testlib/assertlib.py]: Maybe a new line after `"""`?

Comment (EdDev) [reply tests/integration/testlib/assertlib.py]: Done

File: tests/integration/testlib/statelib.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    Report the current state, filtering based on the given interface names.
+    """
+    base_filter_state = {
+        INTERFACES: [{'name': ifname} for ifname in ifnames]
+    }
+    return filter_current_state(base_filter_state)
+
+
+def filter_current_state(desired_state):
```
Comment (tyll) [inline tests/integration/testlib/statelib.py]: Maybe use a name like `get_corresponding_current_state` - it is strange that something that filters the current state gets a desired state as argumentr. Or maybe make the current state an argument as well and use a matching name. Since it get's the current state in the function I would even argue that this should be a class method for a class that represents the current state - but since it is not a class, it is a strange method.

File: tests/integration/static_ip_address_test.py (hunk 6 above)
Comment (EdDev) [reply tests/integration/static_ip_address_test.py]: I just added a second test, hope this is now ok.

File: tests/integration/testlib/statelib.py (hunk 2 above)
Comment (EdDev) [reply tests/integration/testlib/statelib.py]: See #110

**Comment by EdDev** [comment]
ci test please

File: tests/integration/testlib/statelib.py (hunk 2 above)
Comment (tyll) [reply tests/integration/testlib/statelib.py]: Thank you

File: tests/integration/testlib/assertlib.py (hunk 5 above)
Comment (tyll) [reply tests/integration/testlib/assertlib.py]: IMHO normalization does not describe this. For me it means that the state is changed to an equivalent state.

Comment (tyll) [reply tests/integration/testlib/assertlib.py]: Btw. is it expected that this will only work in the docker environment but not with real NICs? I guess then the ethernet subtree will be added and might break the expectations.

Comment (EdDev) [reply tests/integration/testlib/assertlib.py]: I have renamed things in #111 , lets discuss it there.
I am not sure we have a problem with `ethernet`, from "outside" I am not sure the special handling is relevant (there are no None/s and stuff). But lets see how it works out through an added test later on.
